### DISCUSSION TITLE
Issue with getting a Date Value from a field on the Safari Browser on Browserstack

### DISCUSSION THREAD

**Discussion by troyyer** [webdriverio/webdriverio#12878] [Q&A]
Hi there,

I'm unsure if this is related to Browserstack or if it's related to the Safari Browser itself so could I request someone with a MacOS device (I just have Windows machines from Work and personal) to please run this locally on their machine and check if it's working on their machines as well as on a cloud vendor if you have access to one.
I have a weird behaviour I am seeing on Safari where a field that has a Date Value is not being retrieved using `getText()` whereas the same will work with Chrome 
I have saved the code in this repos and it assumes that you have a Browserstack account (potentially it could run on any of the other cloud providers by running the wdio cli install)
https://github.com/troyyer/wdiov8browserstackSafariDateField 
I have a test to fetch some values from a Salesforce page. when i get the text value for a string value, it comes thru without any issue but when the field has a Date Value, it always comes up as blank
I found a Salesforce preview site to trial this out ( it's a very dirty script with a long wait for the iframes to load)
![image](https://github.com/webdriverio/webdriverio/assets/28956675/c931949d-f418-44f2-9a19-a2cde320fc70)
I have highlighted the 2 fields I am trying to `getText()` for
in the code I have a console log 
```
 
  console.log(`browser ${browser.capabilities.browserName} this is a text field: `+ await $(`//c-date/div/h2`).getText())
  console.log(`browser ${browser.capabilities.browserName} this is a Date field: `+ await $(`//p//lightning-formatted-date-time[1]`).getText())
    
```
on Safari the console.log shows the date as blank and the Text field appearing
![image](https://github.com/webdriverio/webdriverio/assets/28956675/a909f05b-87bd-4123-bef2-2d6c6186221d)
![image](https://github.com/webdriverio/webdriverio/assets/28956675/e07cc91e-3d63-4a5b-8a27-cbe658d49775)
On chrome this works
![image](https://github.com/webdriverio/webdriverio/assets/28956675/a7af36b3-e1ed-4b0b-be47-d32fcbfdf87c)
![image](https://github.com/webdriverio/webdriverio/assets/28956675/a1a98828-93ff-4205-895e-a35151fcc373)

I have also trialled with putting in a debug `await browser.debug()` after the iframes load and trialed selecting the data from the REPL

```
 $(`//p//lightning-formatted-date-time[1]`).getText()
 
$(`//c-date/div/h2`).getText()
```
with Chrome 
![image](https://github.com/webdriverio/webdriverio/assets/28956675/aeccc863-77f3-4f45-bcac-59aa475875a3)
with Safari
![image](https://github.com/webdriverio/webdriverio/assets/28956675/c772d5da-d671-4f1a-95fa-87b7364cadaa)
![image](https://github.com/webdriverio/webdriverio/assets/28956675/f882e9d0-503a-4da1-87de-4feb3b714642)

So depending on someone help with some additional findings, I will raise something either in the WebdriverIO repo or will post it to the Browserstack support

**Comment by portyu9** [comment]
Hi @troyyer 

I don't think the evidence is enough to call this a BrowserStack bug.

The key detail in the reproduction is this selector:

```js
await $('//p//lightning-formatted-date-time[1]').getText()
```

`lightning-formatted-date-time` is a Lightning Web Component host. Salesforce base components render their implementation inside **shadow DOM**, so you're asking WebDriver for the text of the custom-element host rather than explicitly locating the rendered content inside its shadow root.

Chrome and Safari drivers have historically differed around how `Get Element Text` behaves across that boundary, which can explain why Chrome returns the rendered date while Safari returns an empty string.

I would first confirm the shadow-DOM boundary:

```js
const date = await $('//p//lightning-formatted-date-time[1]')

console.log(await browser.execute((el) => ({
    textContent: el.textContent,
    innerText: el.innerText,
    hasShadowRoot: Boolean(el.shadowRoot),
    shadowText: el.shadowRoot?.textContent
}), date))
```

If the rendered value is inside an open shadow root, target the actual rendered child with WebdriverIO's `shadow$()` rather than relying on `getText()` on the host:

```js
const host = await $('lightning-formatted-date-time')
const valueElement = await host.shadow$('YOUR_INNER_SELECTOR')
console.log(await valueElement.getText())
```

Use the actual inner selector from the component DOM rather than guessing one.

Current WebdriverIO can automatically traverse shadow roots in BiDi-capable environments, while `shadow$()` remains the explicit solution where that automatic behavior isn't available.

I successfully reviewed the linked reproduction repository and its WDIO config/test. The GitHub-hosted screenshots in the old discussion could no longer be fetched, so I did not rely on them.

Please mark this answer as accepted if it helped, thank you.
